src/VisualStudio/LiveShare/Impl/RunCodeActionsHandler.cs

```diff
@@ -0,0 +1,118 @@
+                    // TODO - This UI thread dependency must be removed.
+                    // https://github.com/dotnet/roslyn/projects/45#card-20619668
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    foreach (var operation in await actionToRun.GetOperationsAsync(cancellationToken).ConfigureAwait(false))
+                    {
+                        operation.Apply(requestContext.Context.Workspace, cancellationToken);
+                    }
+                }
+
+                return true;
+            }
+            else
+            {
+                // We may not have a handler for the command identifier passed in so the base throws.
+                // This is possible, especially in VSCode scenarios, so rather than blow up, log telemetry.
+                Logger.Log(FunctionId.Liveshare_UnknownCodeAction, KeyValueLogMessage.Create(m => m["command"] = request.Command));
+
+                // Return true even in the case that we didn't run the command. Returning false would prompt the guest to tell the host to
+                // enable command executuion, which wouldn't solve their problem.
+                return true;
+            }
+        }
+    }
+
+    [ExportLspRequestHandler(LiveShareConstants.RoslynContractName, LSP.Methods.WorkspaceExecuteCommandName)]
+    [Obsolete("Used for backwards compatibility with old liveshare clients.")]
+    internal class RoslynRunCodeActionsHandler : RunCodeActionsHandler
+    {
+        [ImportingConstructor]
+        public RoslynRunCodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService, IThreadingContext threadingContext) : base(codeFixService, codeRefactoringService, threadingContext)
+        {
+        }
+    }
+
+    [ExportLspRequestHandler(LiveShareConstants.CSharpContractName, LSP.Methods.WorkspaceExecuteCommandName)]
+    internal class CSharpRunCodeActionsHandler : RunCodeActionsHandler
+    {
+        [ImportingConstructor]
+        public CSharpRunCodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService, IThreadingContext threadingContext) : base(codeFixService, codeRefactoringService, threadingContext)
+        {
+        }
+    }
+
+    [ExportLspRequestHandler(LiveShareConstants.VisualBasicContractName, LSP.Methods.WorkspaceExecuteCommandName)]
+    internal class VisualBasicRunCodeActionsHandler : RunCodeActionsHandler
+    {
+        [ImportingConstructor]
+        public VisualBasicRunCodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService, IThreadingContext threadingContext) : base(codeFixService, codeRefactoringService, threadingContext)
+        {
+        }
+    }
+
+    [ExportLspRequestHandler(LiveShareConstants.TypeScriptContractName, LSP.Methods.WorkspaceExecuteCommandName)]
+    internal class TypeScriptRunCodeActionsHandler : RunCodeActionsHandler
```

Comment:
what about things like Xaml?  don't they use Roslyn for code actions?

Reply:
Not through LSP, or at least not yet.

Reply:
Everything that uses the UI thread from a code action is legacy really. In the near future it will be prohibitively expensive.